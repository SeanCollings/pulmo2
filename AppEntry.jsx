import React, { useState, useEffect, useCallback } from 'react';
import * as Font from 'expo-font';
import { View, StyleSheet, Text, ActivityIndicator, Image } from 'react-native';
import Constants from 'expo-constants';
import * as SplashScreen from 'expo-splash-screen';

import Navigator from './navigation';
import ExcerciseContextProvider from './context/excercise-context';
import ProfileContextProvider from './context/profile-context';
import CustomExcerciseContextProvider from './context/custom-excercise-context';
import HistoryContextProvider from './context/history-context';
import SettingsContextProvider from './context/settings-context';
import loadCustomExcercisesAsync from './app-initialise/load-custom-excercise';
import loadSettingsAsync from './app-initialise/load-settings';
import loadProfileAsync from './app-initialise/load-profile';
import loadFavActivityIdArrayAsync from './app-initialise/load-fav-activity-id-array';
import loadActivityIdArrayAsync from './app-initialise/load-activity-id-array';
import AppSetup from './setup';
import SplashImage from './assets/pulmo2_splash.png';

const fetchFonts = () => {
  return Font.loadAsync({
    'tit-light': require('./assets/fonts/TitilliumWeb-Light.ttf'),
    'tit-regular': require('./assets/fonts/TitilliumWeb-Regular.ttf'),
    'tit-bold': require('./assets/fonts/TitilliumWeb-Bold.ttf'),
  });
};

export default function App() {
  const [dataLoaded, setDataLoaded] = useState(false);
  const [setupApp, setSetupApp] = useState(false);
  const [customExcercises, setCustomExcercises] = useState([]);
  const [activityIdArray, setActivityIdArray] = useState([]);
  const [favActivityIdArray, setFavActivityIdArray] = useState([]);
  const [profile, setProfile] = useState({});
  const [appSettings, setAppSettings] = useState({});
  const [showSplashScreen, setShowSplashScreen] = useState(true);

  const updateAsyncState = ({
    customExcercisesAsync,
    activityIdArrayAsync,
    favActivityIdArrayAsync,
    profileAsync,
    settingsAsync,
  }) => {
    customExcercisesAsync && setCustomExcercises(customExcercisesAsync);
    activityIdArrayAsync && setActivityIdArray(activityIdArrayAsync);
    favActivityIdArrayAsync && setFavActivityIdArray(favActivityIdArrayAsync);
    profileAsync && setProfile(profileAsync);
    settingsAsync && setAppSettings(settingsAsync);

    return Promise.resolve();
  };

  const loadAsyncDependencies = async () => {
    try {
      const [
        customExcercisesAsync,
        activityIdArrayAsync,
        favActivityIdArrayAsync,
        profileAsync,
        settingsAsync,
      ] = await Promise.all([
        loadCustomExcercisesAsync(),
        loadActivityIdArrayAsync(),
        loadFavActivityIdArrayAsync(),
        loadProfileAsync(),
        loadSettingsAsync(),
      ]);

      return await updateAsyncState({
        customExcercisesAsync,
        activityIdArrayAsync,
        favActivityIdArrayAsync,
        profileAsync,
        settingsAsync,
      });
    } catch (err) {
      throw new Error(err);
    }
  };

  const prepareResources = useCallback(async () => {
    try {
      await Promise.all([fetchFonts(), loadAsyncDependencies()]);

      SplashScreen.hideAsync().then(() => {
        setDataLoaded(true);
        setTimeout(() => {
          setShowSplashScreen(false);
        }, 1000);
      });
    } catch (err) {
      console.log(`prepareResources error: ${err}`);
    }
  }, []);

  useEffect(() => {
    if (Constants.manifest?.extra?.setupData === 'true') {
      setSetupApp(true);
    } else {
      Text.defaultProps = Text.defaultProps || {};
      Text.defaultProps.allowFontScaling = false;

      prepareResources();
    }
  }, []);

  if (setupApp) return <AppSetup />;

  return (
    <View style={styles.container} data-testid="app-component">
      {showSplashScreen && (
        <View style={styles.splashScreenContainer}>
          <Image
            style={styles.image}
            source={SplashImage}
            resizeMode="contain"
          />
          <ActivityIndicator
            size="large"
            style={styles.indicator}
            color="#fff"
          ></ActivityIndicator>
        </View>
      )}
      {dataLoaded && (
        <CustomExcerciseContextProvider excercises={customExcercises}>
          <ExcerciseContextProvider>
            <ProfileContextProvider profile={profile}>
              <HistoryContextProvider
                idArray={activityIdArray}
                favIdArray={favActivityIdArray}
              >
                <SettingsContextProvider settings={appSettings}>
                  <Navigator />
                </SettingsContextProvider>
              </HistoryContextProvider>
            </ProfileContextProvider>
          </ExcerciseContextProvider>
        </CustomExcerciseContextProvider>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#002f56' },
  splashScreenContainer: {
    backgroundColor: '#002f56',
    justifyContent: 'center',
    zIndex: 1,
    position: 'absolute',
    height: '100%',
    width: '100%',
  },
  image: {
    width: '100%',
    height: '100%',
  },
  indicator: {
    position: 'absolute',
    alignSelf: 'center',
    opacity: 0.7,
  },
});
