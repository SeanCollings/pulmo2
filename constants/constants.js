export const COLORS = {
  PRIMARY: '#005bbb',
  SECONDARY: '#002f56',
  TERTIARY: '#00b8bb',
  QUARTERNARY: '#5bacf3',
  QUINERY: '#990000',
  TEXT: '#3e4b57',
  BACKGROUND: 'white',
  SECONDARY_TEXT: 'white',
  BORDER: '#CCC',
  ERROR: '#ff1507',
};
export const COLORS_DARK = {
  DARK: true,
  PRIMARY: '#222222',
  SECONDARY: '#2f9fd0',
  TERTIARY: '#28D6C0',
  QUARTERNARY: '#ff895e',
  QUINERY: '#990000',
  TEXT: '#ededed',
  BACKGROUND: '#121212',
  SECONDARY_TEXT: '#ededed',
  BORDER: '#afafaf',
  ERROR: '#fb3c36',
  DARK_GRAY: '#707070',
  DARKER_GRAY: '#2d2d2d',
};
// export const COLORS_DARK = {
//   PRIMARY: '#222222',
//   SECONDARY: '#2f9fd0',
//   TERTIARY: '#28D6C0',
//   QUARTERNARY: '#F12B7E',
//   QUINERY: '#990000',
//   TEXT: 'white',
//   BACKGROUND: '#121212',
//   SECONDARY_TEXT: 'white',
//   BORDER: 'white',
// };

// export const COLORS_DARK = {
//   PRIMARY: '#07121B',
//   SECONDARY: '#28D6C0',
//   TERTIARY: '#F12B7E',
//   TEXT: '#CCC',
//   BACKGROUND: '#07121B',
//   SECONDARY_TEXT: '#6e6e6e',
// };

export const APP_ID = '@Plumo2:';
export const TOTAL_DIFFICULTY_LEVELS = 5;
export const LIGHT_MODE = 'light_mode';
export const DARK_MODE = 'dark_mode';

export const INTIAL_STATE = 'initial';
export const BREATHING_STATE = 'breathing';
export const REST_STATE = 'rest';
export const END_STATE = 'end';

export const ROUND_STATES = [
  INTIAL_STATE,
  BREATHING_STATE,
  REST_STATE,
  END_STATE,
];

export const INSTRUCTIONS = {
  [INTIAL_STATE]: {
    state: INTIAL_STATE,
    main: '',
    sub: `rounds total`,
    prompt: 'tap to begin',
  },
  [BREATHING_STATE]: {
    state: BREATHING_STATE,
    main: '',
    sub: '',
    prompt: ' Inhale & Exhale',
  },
  [REST_STATE]: {
    state: REST_STATE,
    main: 'Rest period',
    sub: '',
    prompt: '',
  },
  [END_STATE]: {
    state: END_STATE,
    main: 'Your training is complete',
    sub: '',
    prompt: 'Complete',
  },
};
