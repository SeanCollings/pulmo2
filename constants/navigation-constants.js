import { Platform } from 'react-native';

export const defaultNavOptions = (theme) => ({
  headerStyle: {
    backgroundColor: theme.PRIMARY,
    borderBottomWidth: 1,
    borderBottomColor: theme.BORDER,
    shadowColor: 'transparent',
    shadowOffset: 0,
    shadowOpacity: 0,
  },
  headerTitleStyle: {
    fontFamily: 'tit-regular',
    backgroundColor:
      Platform.OS === 'android' ? theme.PRIMARY : theme.SECONDARY_TEXT,
  },
  headerTintColor:
    Platform.OS === 'android' ? theme.SECONDARY_TEXT : theme.PRIMARY,
});
