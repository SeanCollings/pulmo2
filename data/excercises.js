import { DATA, STRENGTH_KEY, ENDURANCE_KEY, CUSTOM_KEY } from '.';

const STRENGTH = {
  id: STRENGTH_KEY,
  title: STRENGTH_KEY,
  icon: 'dumbbell',
  contents: DATA[STRENGTH_KEY],
};
const ENDURANCE = {
  id: ENDURANCE_KEY,
  title: ENDURANCE_KEY,
  icon: 'clock',
  contents: DATA[ENDURANCE_KEY],
};
const CUSTOM = {
  id: CUSTOM_KEY,
  title: CUSTOM_KEY,
  icon: 'wrench',
  contents: DATA[CUSTOM_KEY],
};

export const EXCERCISE_DATA = [STRENGTH, ENDURANCE, CUSTOM];
