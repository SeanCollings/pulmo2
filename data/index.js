export const STRENGTH_KEY = 'Strength';
export const ENDURANCE_KEY = 'Endurance';
export const CUSTOM_KEY = 'Custom';

export const DATA = {
  [STRENGTH_KEY]: [
    {
      id: 11,
      title: 'Strength Level I',
      cycles: 10,
      rest: 60,
      rounds: 4,
    },
    {
      id: 12,
      title: 'Strength Level II',
      cycles: 10,
      rest: 60,
      rounds: 5,
    },
    {
      id: 13,
      title: 'Strength Level III',
      cycles: 12,
      rest: 50,
      rounds: 5,
    },
  ],
  [ENDURANCE_KEY]: [
    {
      id: 21,
      title: 'Endurance Level I',
      cycles: 10,
      rest: 30,
      rounds: 4,
    },
    {
      id: 22,
      title: 'Endurance Level II',
      cycles: 12,
      rest: 20,
      rounds: 5,
    },
    {
      id: 23,
      title: 'Endurance Level III',
      cycles: 15,
      rest: 20,
      rounds: 5,
    },
  ],
  [CUSTOM_KEY]: [],
};

export const DEFAULT_EXCERCISE = DATA[STRENGTH_KEY][0];
