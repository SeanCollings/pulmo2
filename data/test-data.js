import { THEME } from '../context/settings-context';
import { DARK_MODE } from '../constants/constants';
import { SELECTED_LEVEL } from '../context/profile-context';
import { CUSTOM_TAG } from '../context/custom-excercise-context';
import { ACTIVITY_TAG, FAV_ACTIVITY_TAG } from '../context/history-context';
import { SELECTED_EXCERCISE } from '../context/excercise-context';

export const TEST_ALL_ACTIVITIES = 'test_all_activities';
export const TEST_KEY = 'test_key';
export const TEST_DATA = 'test_data';
export const TEST_TITLE = 'test_title';
export const TEST_MUTLIPLE = 'test_multiple';
export const TEST_MUTLIPLE_ID = 'test_multiple_id';
export const TEST_PRE_ID = 'test_pre_id';

//#region Custom Excercise ID Array
const test_customExcerciseIdArray = {
  [TEST_TITLE]: 'Custom Excercise IDs',
  [TEST_KEY]: CUSTOM_TAG,
  [TEST_DATA]: [1597490106660, 1597490120213],
};
//#endregion Custom Excercises ID Array

//#region Custom Excercises
const test_customExcercises = {
  [TEST_TITLE]: 'Custom Excercises',
  [TEST_MUTLIPLE_ID]: 'id',
  [TEST_PRE_ID]: CUSTOM_TAG,
  [TEST_DATA]: [
    {
      cycles: '12',
      id: 1597490106660,
      rest: '3',
      rounds: '4',
      title: 'Custom Excercise',
    },
    {
      cycles: '10',
      id: 1597490120213,
      rest: '1',
      rounds: '10',
      title: 'Many Rounds',
    },
  ],
};
//#endregion Custom Excercises

//#region Selected Excercise
const test_selected_excercise = {
  [TEST_TITLE]: 'Selected Excercise',
  [TEST_KEY]: SELECTED_EXCERCISE,
  [TEST_DATA]: [1597490106660, 'Custom'],
};
//#endregion Selected Excercise

//#region Activity ID Array
const test_activityIdArray = {
  [TEST_TITLE]: 'Activity IDs',
  [TEST_KEY]: ACTIVITY_TAG,
  [TEST_DATA]: [
    '2020-09-01T12:50:14.562Z',
    '2020-08-25T10:12:47.005Z',
    '2020-08-20T13:24:38.330Z',
    '2020-08-20T10:53:57.797Z',
    '2020-08-19T15:14:17.865Z',
    '2020-08-15T12:52:43.310Z',
    '2020-08-12T12:12:12.100Z',
    '2020-08-10T09:51:44.609Z',
    '2020-08-09T12:50:58.499Z',
    '2020-08-08T11:36:50.680Z',
    '2020-08-07T12:55:11.988Z',
    '2020-08-05T12:44:47.027Z',
    '2020-08-01T12:42:37.786Z',
    '2020-07-18T16:41:15.186Z',
    '2020-07-17T12:40:54.141Z',
    '2020-07-15T12:38:25.328Z',
    '2020-07-12T10:47:28.181Z',
    '2020-07-09T12:36:46.328Z',
    '2020-07-09T10:26:11.702Z',
    '2020-06-10T12:35:09.722Z',
    '2020-06-10T11:19:55.838Z',
  ],
};
//#endregion Activity ID Array

//#region Fav Activity ID Array
const test_favActivityIdArray = {
  [TEST_TITLE]: 'Favourite Activity IDs',
  [TEST_KEY]: FAV_ACTIVITY_TAG,
  [TEST_DATA]: [
    '2020-08-20T13:24:38.330Z',
    '2020-08-19T15:14:17.865Z',
    '2020-08-10T09:51:44.609Z',
    '2020-08-09T12:50:58.499Z',
    '2020-08-05T12:44:47.027Z',
    '2020-07-18T16:41:15.186Z',
    '2020-07-17T12:40:54.141Z',
    '2020-07-09T12:36:46.328Z',
    '2020-06-10T11:19:55.838Z',
  ],
};
//#endregion Fav Activity ID Array

//#region Profile
const test_profile_selected_level = {
  [TEST_TITLE]: 'Profile',
  [TEST_KEY]: SELECTED_LEVEL,
  [TEST_DATA]: 2,
};
//#endregion Profile

//#region Settings
const test_settings_theme = {
  [TEST_TITLE]: 'Settings',
  [TEST_KEY]: THEME,
  [TEST_DATA]: DARK_MODE,
};
//#endregion Settings

//#region All Activities
const test_all_activities = {
  [TEST_TITLE]: 'All Activities',
  [TEST_MUTLIPLE_ID]: 'date',
  [TEST_KEY]: TEST_ALL_ACTIVITIES,
  [TEST_DATA]: [
    {
      date: '2020-09-01T12:50:14.562Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise Here',
      },
      favourite: false,
      level: 2,
      results: [
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-25T10:12:47.005Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2,
      results: [
        ['00:03', '00:03'],
        ['00:03', '00:03'],
        ['00:03', '00:03'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-20T13:24:38.330Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise 1',
      },
      favourite: true,
      level: 2,
      rating: 4,
      results: [
        ['00:07', '00:03'],
        ['00:02', '00:03'],
        ['00:07', '00:03'],
        ['00:04', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-20T10:53:57.797Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2,
      results: [
        ['00:04', '00:03'],
        ['00:07', '00:03'],
        ['00:03', '00:03'],
        ['00:08', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-19T15:14:17.865Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      favourite: true,
      level: 2,
      results: [
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-15T12:52:43.310Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 2.5,
      results: [['00:09', '00:01'], ['00:09', '00:01'], ['00:09', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-12T12:12:12.100Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 2.5,
      results: [['00:02', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-10T09:51:44.609Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      favourite: true,
      incomplete: ['feeling_lazy'],
      level: 2.5,
      results: [['00:02', '00:01'], ['00:02', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-09T12:50:58.499Z',
      excercise: {
        cycles: 12,
        id: 13,
        rest: 50,
        rounds: 5,
        title: 'Strength Level III',
      },
      favourite: true,
      level: 2.5,
      results: [
        ['00:05', '00:50'],
        ['00:05', '00:50'],
        ['00:06', '00:50'],
        ['00:06', '00:50'],
        ['00:04', ''],
      ],
      type: 'Strength',
    },
    {
      date: '2020-08-08T11:36:50.680Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2.5,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-07T12:55:11.988Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-05T12:44:47.027Z',
      excercise: {
        cycles: 12,
        id: 22,
        rest: 20,
        rounds: 5,
        title: 'Endurance Level II',
      },
      favourite: true,
      level: 3,
      results: [
        ['00:03', '00:20'],
        ['00:04', '00:20'],
        ['00:08', '00:20'],
        ['00:05', '00:20'],
        ['00:05', ''],
      ],
      type: 'Endurance',
    },
    {
      date: '2020-08-01T12:42:37.786Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3,
      results: [
        ['00:04', '00:03'],
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-18T16:41:15.186Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Custom Many Rounds',
      },
      favourite: true,
      incomplete: ['out_of_breath'],
      level: 2.5,
      rating: 3,
      results: [['00:01', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-17T12:40:54.141Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      favourite: true,
      incomplete: ['other'],
      level: 3,
      results: [['00:08', '00:01'], ['00:05', '00:01'], ['00:05', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-15T12:38:25.328Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds 2',
      },
      level: 3,
      results: [
        ['00:02', '00:30'],
        ['00:02', '00:30'],
        ['00:03', '00:30'],
        ['00:03', '00:30'],
        ['00:04', '00:30'],
        ['00:04', '00:30'],
        ['00:03', '00:30'],
        ['00:02', '00:30'],
        ['00:03', '00:30'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-12T10:47:28.181Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 3,
      results: [['00:03', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-09T12:36:46.328Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds Edited',
      },
      favourite: true,
      incomplete: ['placeholder'],
      level: 2.5,
      rating: 2,
      results: [
        ['00:03', '00:50'],
        ['00:04', '00:50'],
        ['00:05', '00:50'],
        ['00:02', '00:50'],
        [],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-09T10:26:11.702Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['feeling_lazy'],
      level: 3.5,
      results: [['00:02', '00:01'], ['00:04', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-06-10T12:35:09.722Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3.5,
      results: [
        ['00:01', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-06-10T11:19:55.838Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      favourite: true,
      level: 3.5,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:03', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
  ],
};
//#endregion All Activities

export default {
  test_customExcerciseIdArray,
  test_customExcercises,
  test_selected_excercise,
  test_activityIdArray,
  test_favActivityIdArray,
  test_profile_selected_level,
  test_settings_theme,
  test_all_activities,
};
