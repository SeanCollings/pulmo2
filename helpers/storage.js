import { AsyncStorage } from 'react-native';
import { APP_ID } from '../constants/constants';

export const clearAllAsyncStorage = () => {
  AsyncStorage.clear();
};

export const getAsyncData = async (id) => {
  try {
    const value = await AsyncStorage.getItem(`${APP_ID}${id}`);
    if (value !== null) {
      return JSON.parse(value);
    } else {
      return null;
    }
  } catch (error) {
    console.log('Failure retrieving ' + id + ' error: ' + error);
  }
};

export const getMultiAsyncData = async (id, keys) => {
  try {
    const formattedKeys = keys.map((key) => `${APP_ID}${id}${key}`);
    const values = await AsyncStorage.multiGet(formattedKeys);

    if (values && !!values.length) return values;
    else {
      return null;
    }
  } catch (err) {
    console.log('Failure retrieving from all keys with error ' + error);
  }
};

export const storeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.setItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    console.log('Failure saving: ' + error);
    return false;
  }
};

export const removeAsyncData = async (id) => {
  try {
    await AsyncStorage.removeItem(`${APP_ID}${id}`);
  } catch (err) {
    console.log('Failure removing: ' + error);
  }
};
