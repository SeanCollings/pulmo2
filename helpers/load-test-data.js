import testData, {
  TEST_KEY,
  TEST_TITLE,
  TEST_MUTLIPLE_ID,
  TEST_DATA,
  TEST_PRE_ID,
} from '../data/test-data';
import { getAsyncData, storeAsyncData } from './storage';

export const MESSAGE_UNKNOWN = 'unknown';
export const MESSAGE_PRE_SAVED = 'pre-saved';
export const MESSAGE_DONE = 'done';

export default async (title) => {
  try {
    const selectedData = Object.values(testData).find(
      (data) => data[TEST_TITLE] === title
    );

    if (!selectedData) return MESSAGE_UNKNOWN;

    if (!selectedData[TEST_MUTLIPLE_ID]) {
      const savedData = await getAsyncData(selectedData[TEST_KEY]);

      if (!savedData) {
        console.log('Saving data for:', selectedData[TEST_KEY]);
        await storeAsyncData(selectedData[TEST_KEY], selectedData[TEST_DATA]);
        return MESSAGE_DONE;
      } else {
        return MESSAGE_PRE_SAVED;
      }
    } else {
      const updatedSaves = await selectedData[TEST_DATA].reduce(
        async (acc, data) => {
          const newAcc = await acc;
          const id = selectedData[TEST_MUTLIPLE_ID];
          const preId = selectedData[TEST_PRE_ID] || '';
          const savedData = await getAsyncData(`${preId}${data[id]}`);

          if (!savedData) {
            console.log('Saving data for:', `${preId}${data[id]}`);
            newAcc.push(data[id]);
            await storeAsyncData(`${preId}${data[id]}`, data);
          }
          return newAcc;
        },
        []
      );

      if (!updatedSaves.length) return MESSAGE_PRE_SAVED;

      return `${MESSAGE_DONE} [${updatedSaves.length}]`;
    }
  } catch (err) {
    throw new Error(err);
  }
};
