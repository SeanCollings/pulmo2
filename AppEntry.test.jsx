import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup } from 'react-native-testing-library';

import AppEntry from './AppEntry';
import fixtures from './app-initialise/__fixtures__/index.fixtures';
import loadCustomExcercisesAsync from './app-initialise/load-custom-excercise';
import loadActivityIdArrayAsync from './app-initialise/load-activity-id-array';
import loadFavActivityIdArrayAsync from './app-initialise/load-fav-activity-id-array';
import loadProfileAsync from './app-initialise/load-profile';
import loadSettingsAsync from './app-initialise/load-settings';
import { LIGHT_MODE, DARK_MODE } from './constants/constants';
import { getAsyncData } from './helpers/storage';
import { STRENGTH_KEY } from './data';

jest.useFakeTimers();
jest.mock('Dimensions');
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('./app-initialise/load-activity-id-array');
jest.mock('./app-initialise/load-custom-excercise');
jest.mock('./app-initialise/load-fav-activity-id-array');
jest.mock('./app-initialise/load-profile');
jest.mock('./app-initialise/load-settings');
jest.mock('./helpers/storage');

describe('AppEntry', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  describe('success', () => {
    test('has 1 child', async () => {
      let tree;
      getAsyncData.mockReturnValueOnce([11, STRENGTH_KEY]);

      await act(async () => {
        tree = create(<AppEntry />);
      });
      expect(tree.toJSON().children.length).toBe(1);
    });

    // Set to test.skip as Azure produces a different snapshot for some reason.
    // Remove skip to test App workflow locally.
    test.skip('renders in light mode', async () => {
      let tree;
      loadActivityIdArrayAsync.mockReturnValueOnce(fixtures.activityIdArray);
      loadCustomExcercisesAsync.mockReturnValueOnce(fixtures.customExcercise);
      loadFavActivityIdArrayAsync.mockReturnValueOnce(
        fixtures.favActivityIdArray
      );
      loadProfileAsync.mockReturnValueOnce(fixtures.profile);
      loadSettingsAsync.mockReturnValueOnce({
        ...fixtures.settings,
        theme: LIGHT_MODE,
      });
      await act(async () => {
        tree = create(<AppEntry />);
      });
      expect(tree).toMatchSnapshot();
    });

    // Set to test.skip as Azure produces a different snapshot for some reason.
    // Remove skip to test App workflow locally.
    test.skip('renders in dark mode', async () => {
      let tree;
      loadSettingsAsync.mockReturnValueOnce({
        ...fixtures.settings,
        theme: DARK_MODE,
      });
      await act(async () => {
        tree = create(<AppEntry />);
      });
      expect(tree).toMatchSnapshot();
    });
  });

  describe('errors', () => {
    test('logs errors', async () => {
      console.log = jest.fn();
      loadActivityIdArrayAsync.mockReturnValueOnce(null);
      loadCustomExcercisesAsync.mockReturnValueOnce(
        new Promise((_, reject) => {
          reject(new Error('reject'));
        })
      );
      loadFavActivityIdArrayAsync.mockReturnValueOnce(null);
      loadProfileAsync.mockReturnValueOnce(null);
      loadSettingsAsync.mockReturnValueOnce(null);

      await act(async () => {
        create(<AppEntry />);
        jest.runAllTimers();
      });
      expect(console.log).toHaveBeenCalled();
    });
  });
});
