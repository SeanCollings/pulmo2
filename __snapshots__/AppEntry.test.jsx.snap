// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppEntry errors should show the splash screen on error 1`] = `
<View
  data-testid="app-component"
  style={
    Object {
      "backgroundColor": "#002f56",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#002f56",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={1}
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
    <ActivityIndicator
      animating={true}
      color="#fff"
      hidesWhenStopped={true}
      size="large"
      style={
        Object {
          "alignSelf": "center",
          "opacity": 0.7,
          "position": "absolute",
        }
      }
    />
  </View>
</View>
`;
